app/models/concerns/has_locking.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+# This concern provides functionality for locking items along with additional functionality to handle
+# the concept of retrying items after a certain period of time. The following database columns are
+# required on the model
+#
+# * locked_by - A string column to store the name of the process that has locked the item
+# * locked_at - A datetime column to store the time the item was locked
+# * retry_after - A datetime column to store the time after which the item should be retried
+# * attempts - An integer column to store the number of attempts that have been made to process the item
+#
+module HasLocking
+
+  extend ActiveSupport::Concern
+
+  included do
+    scope :unlocked, -> { where(locked_at: nil) }
+    scope :queueable, -> { where("retry_after IS NULL OR retry_after < ?", Time.now) }
+    scope :retriable, -> { where("retry_after IS NULL OR retry_after < ?", Time.now) }
```

Comment:
This redundancy seems unnecessary. I assume a single scope "retriable" would be sufficient.